library/std/src/sys/unix/locks/futex_rwlock.rs

```diff
@@ -0,0 +1,293 @@
+use crate::sync::atomic::{
+    AtomicI32,
+    Ordering::{Acquire, Relaxed, Release},
+};
+use crate::sys::futex::{futex_wait, futex_wake, futex_wake_all};
+
+pub type MovableRwLock = RwLock;
+
+pub struct RwLock {
+    // The state consists of a 30-bit reader counter, a 'readers waiting' flag, and a 'writers waiting' flag.
+    // Bits 0..30:
+    //   0: Unlocked
+    //   1..=0x3FFF_FFFE: Locked by N readers
+    //   0x3FFF_FFFF: Write locked
+    // Bit 30: Readers are waiting on this futex.
+    // Bit 31: Writers are waiting on the writer_notify futex.
+    state: AtomicI32,
+    // The 'condition variable' to notify writers through.
+    // Incremented on every signal.
+    writer_notify: AtomicI32,
+}
+
+const READ_LOCKED: i32 = 1;
+const MASK: i32 = (1 << 30) - 1;
+const WRITE_LOCKED: i32 = MASK;
```

Comment:
It seems a bit strange to represent `WRITE_LOCKED` with a special reader count. I think using a separate bit for `WRITE_LOCKED` would make the code more efficient (e.g. using native overflow flags to detect reader count overflow, using `fetch_or` for locking instead of a CAS loop, etc).

Reply:
I'll try that out and see if it simplifies things. It probably does. Thanks. :)

Reply:
I tried it, but it does not seem to simplify things.

`fetch_or` doesn't seem very useful, because it should only set the bit if the lock isn't read-locked yet. If we unconditionally set the bit when attempting to lock, it will also be set when it's still read-locked and read-unlocking will need special handling for that situation.

Simplifying the overflow check doesn't help, because at the points where this is checked, we also need to check the waiting bits too.

I also prefer to combine the fields, to make invalid states unrepresentable. (Effectively `enum { Unlocked, ReadLocked(NonZeroU30), WriteLocked }` instead of `struct { read_locked: u30, write_locked: bool }`.)